File: Filesystem.py
import os
import sys

"""
Simple File System Simulator 
Disk Layout:
  - 100 blocks total
  - Blocks  0–9  : Free Map  (10 blocks reserved)
  - Blocks 10–99 : File storage (90 blocks)

Individual Block Layout (544 bytes each):
  - Bytes   0–31  : File name  (32 bytes, null-padded)
  - Bytes  32–543 : File data  (512 bytes, null-padded)

Free Map:
  - Stored as the first 100 bytes of the Free Map region on disk
  - 0 = free block, 1 = allocated block
  - Blocks 0–9 are always marked allocated (reserved for Free Map)

DONE  : Disk layout, block I/O, free map (load/save/scan), format, REPL shell
Not Done: Rebuild file_table in startup(), then implement create, delete, ls
Not Done: Implement read, write, and full error handling
"""

# Constants
DISK_FILE        = "disk.bin"
TOTAL_BLOCKS     = 100
FREE_MAP_BLOCKS  = 10          # Blocks 0–9 Reserved
FILE_START_BLOCK = 10          # First Usable File Block
FILENAME_SIZE    = 32          # Bytes Per Filename Field
DATA_SIZE        = 512         # Bytes Per Data Field
BLOCK_SIZE       = FILENAME_SIZE + DATA_SIZE   # 544 Bytes Per Block
DISK_SIZE        = TOTAL_BLOCKS * BLOCK_SIZE   # 54,400 Bytes Total


# Low-level Disk I/O
def read_block(block_num: int) -> bytes:
    """Read and return the raw bytes of a single block from disk."""
    with open(DISK_FILE, "rb") as f:
        f.seek(block_num * BLOCK_SIZE)
        return f.read(BLOCK_SIZE)


def write_block(block_num: int, data: bytes) -> None:
    """Write exactly BLOCK_SIZE bytes to a block. Zero-pads if data is short."""
    padded = data[:BLOCK_SIZE].ljust(BLOCK_SIZE, b"\x00")
    with open(DISK_FILE, "r+b") as f:
        f.seek(block_num * BLOCK_SIZE)
        f.write(padded)


# Free Map
def load_free_map() -> list:
    """
    Read the free map from disk.
    Returns a list of 100 ints — one per block (0 = free, 1 = allocated).
    The map is stored as the first TOTAL_BLOCKS bytes of the disk.
    """
    with open(DISK_FILE, "rb") as f:
        f.seek(0)
        raw = f.read(TOTAL_BLOCKS)
    return list(raw)


def save_free_map(free_map: list) -> None:
    """Persist the free map to the reserved region at the start of disk."""
    data = bytes(free_map).ljust(FREE_MAP_BLOCKS * BLOCK_SIZE, b"\x00")
    with open(DISK_FILE, "r+b") as f:
        f.seek(0)
        f.write(data[:FREE_MAP_BLOCKS * BLOCK_SIZE])


def first_free_block(free_map: list) -> int | None:
    """
    Scan the free map and return the lowest-numbered available file block.
    Returns None if the disk is full.
    """
    for i in range(FILE_START_BLOCK, TOTAL_BLOCKS):
        if free_map[i] == 0:
            return i
    return None


# Format Command
def cmd_format(free_map: list, file_table: dict) -> tuple:
    """
    Wipe the entire disk and re-initialize the free map.
    Blocks 0–9 are marked allocated (Free Map region); all others are free.
    """
    print("Formatting disk...")

    with open(DISK_FILE, "wb") as f:
        f.write(b"\x00" * DISK_SIZE)

    new_map = [1] * FREE_MAP_BLOCKS + [0] * (TOTAL_BLOCKS - FREE_MAP_BLOCKS)
    save_free_map(new_map)

    print("Disk formatted successfully.")
    print(f"FreeMap blocks 0-{FREE_MAP_BLOCKS - 1} are now allocated.")

    free_map[:] = new_map
    file_table.clear()

    return free_map, file_table


# Startup
def startup() -> tuple:
    """
    On first run: create a blank disk image.
    On subsequent runs: load the existing disk and rebuild state.
    Returns (free_map, file_table).
    """
    if not os.path.isfile(DISK_FILE) or os.path.getsize(DISK_FILE) != DISK_SIZE:
        with open(DISK_FILE, "wb") as f:
            f.write(b"\x00" * DISK_SIZE)
        free_map  = [0] * TOTAL_BLOCKS
        file_table = {}
        print("No disk image found. A blank disk has been created.")
        print("Tip: run 'format' to initialize the file system before use.")
    else:
        free_map = load_free_map()
        print("Loading file table from disk...")
        file_table = {}  # TODO: Rebuild file table from allocated blocks
        print("File table loaded successfully.")

    return free_map, file_table


# Main REPL
HELP_TEXT = """\
Available commands:
  --> format
  --> create <filename>    
  --> read <filename>       
  --> write <filename>      
  --> delete <filename>      
  --> ls                     
  --> exit"""


def main():
    free_map, file_table = startup()
    print("\nWelcome to the simple file system simulator.")
    print(HELP_TEXT)

    while True:
        try:
            raw = input("\n> ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\nExiting file system simulator.")
            sys.exit(0)

        if not raw:
            continue

        parts   = raw.split(maxsplit=1)
        command = parts[0].lower()
        argument = parts[1].strip() if len(parts) > 1 else ""

        if command == "format":
            free_map, file_table = cmd_format(free_map, file_table)

        elif command in ("create", "read", "write", "delete", "ls"):
            # TODO: Implement these commands
            print(f"'{command}' is not yet implemented.")

        elif command == "exit":
            print("Exiting file system simulator.")
            sys.exit(0)

        else:
            print(f"Unknown command: '{command}'.")


if __name__ == "__main__":
    main()

File: FileSystem.java
import java.io.*;
import java.nio.file.*;
import java.util.*;

/*
 * Simple File System Simulator
 * Disk Layout:
 *   - 100 blocks total
 *   - Blocks  0–9  : Free Map  (10 blocks reserved)
 *   - Blocks 10–99 : File storage (90 blocks)
 *
 * Individual Block Layout (544 bytes each):
 *   - Bytes   0–31  : File name  (32 bytes, null-padded)
 *   - Bytes  32–543 : File data  (512 bytes, null-padded)
 *
 * Free Map:
 *   - Stored as the first 100 bytes of the Free Map region on disk
 *   - 0 = free block, 1 = allocated block
 *   - Blocks 0–9 are always marked allocated (reserved for Free Map)
 *
 * DONE  : Disk layout, block I/O, free map (load/save/scan), format, REPL shell
 * Not Done: Rebuild file_table in startup(), then implement create, delete, ls
 * Not Done: Implement read, write, and full error handling
 */
public class FileSystem {

    static final String DISK_FILE        = "disk.bin";
    static final int    TOTAL_BLOCKS     = 100;
    static final int    FREE_MAP_BLOCKS  = 10;
    static final int    FILE_START_BLOCK = 10;
    static final int    FILENAME_SIZE    = 32;
    static final int    DATA_SIZE        = 512;
    static final int    BLOCK_SIZE       = FILENAME_SIZE + DATA_SIZE;
    static final int    DISK_SIZE        = TOTAL_BLOCKS * BLOCK_SIZE;

    // Low-level Disk I/O
    static byte[] readBlock(int blockNum) throws IOException {
        try (RandomAccessFile raf = new RandomAccessFile(DISK_FILE, "r")) {
            raf.seek((long) blockNum * BLOCK_SIZE);
            byte[] buffer = new byte[BLOCK_SIZE];
            raf.readFully(buffer);
            return buffer;
        }
    }

    static void writeBlock(int blockNum, byte[] data) throws IOException {
        byte[] padded = new byte[BLOCK_SIZE];
        System.arraycopy(data, 0, padded, 0, Math.min(data.length, BLOCK_SIZE));
        try (RandomAccessFile raf = new RandomAccessFile(DISK_FILE, "rw")) {
            raf.seek((long) blockNum * BLOCK_SIZE);
            raf.write(padded);
        }
    }

    // Free Map
    static int[] loadFreeMap() throws IOException {
        try (RandomAccessFile raf = new RandomAccessFile(DISK_FILE, "r")) {
            raf.seek(0);
            byte[] raw = new byte[TOTAL_BLOCKS];
            raf.readFully(raw);
            int[] freeMap = new int[TOTAL_BLOCKS];
            for (int i = 0; i < TOTAL_BLOCKS; i++) freeMap[i] = raw[i] & 0xFF;
            return freeMap;
        }
    }

    static void saveFreeMap(int[] freeMap) throws IOException {
        byte[] data = new byte[FREE_MAP_BLOCKS * BLOCK_SIZE];
        for (int i = 0; i < TOTAL_BLOCKS; i++) data[i] = (byte) freeMap[i];
        try (RandomAccessFile raf = new RandomAccessFile(DISK_FILE, "rw")) {
            raf.seek(0);
            raf.write(data);
        }
    }

    static int firstFreeBlock(int[] freeMap) {
        for (int i = FILE_START_BLOCK; i < TOTAL_BLOCKS; i++)
            if (freeMap[i] == 0) return i;
        return -1;
    }

    // Format Command
    static void cmdFormat(int[] freeMap, Map<String, Integer> fileTable) throws IOException {
        System.out.println("Formatting disk...");
        try (FileOutputStream fos = new FileOutputStream(DISK_FILE)) {
            fos.write(new byte[DISK_SIZE]);
        }
        int[] newMap = new int[TOTAL_BLOCKS];
        for (int i = 0; i < FREE_MAP_BLOCKS; i++) newMap[i] = 1;
        saveFreeMap(newMap);
        System.out.println("Disk formatted successfully.");
        System.out.printf("FreeMap blocks 0-%d are now allocated.%n", FREE_MAP_BLOCKS - 1);
        System.arraycopy(newMap, 0, freeMap, 0, TOTAL_BLOCKS);
        fileTable.clear();
    }

    // Startup
    static class StartupResult {
        int[] freeMap;
        Map<String, Integer> fileTable;
        StartupResult(int[] freeMap, Map<String, Integer> fileTable) {
            this.freeMap = freeMap;
            this.fileTable = fileTable;
        }
    }

    static StartupResult startup() throws IOException {
        File diskFile = new File(DISK_FILE);
        if (!diskFile.exists() || diskFile.length() != DISK_SIZE) {
            try (FileOutputStream fos = new FileOutputStream(DISK_FILE)) {
                fos.write(new byte[DISK_SIZE]);
            }
            System.out.println("No disk image found. A blank disk has been created.");
            System.out.println("Tip: run 'format' to initialize the file system before use.");
            return new StartupResult(new int[TOTAL_BLOCKS], new HashMap<>());
        } else {
            int[] freeMap = loadFreeMap();
            System.out.println("Loading file table from disk...");
            Map<String, Integer> fileTable = new HashMap<>(); // TODO: Rebuild file table
            System.out.println("File table loaded successfully.");
            return new StartupResult(freeMap, fileTable);
        }
    }

    // Main REPL
    static final String HELP_TEXT =
            "Available commands:\n" +
                    "  format\n" +
                    "  create <filename>\n" +
                    "  read <filename>\n" +
                    "  write <filename>\n" +
                    "  delete <filename>\n" +
                    "  ls\n" +
                    "  exit";

    public static void main(String[] args) throws IOException {
        StartupResult state = startup();
        int[] freeMap = state.freeMap;
        Map<String, Integer> fileTable = state.fileTable;

        System.out.println("\nWelcome to the simple file system simulator.");
        System.out.println(HELP_TEXT);

        Scanner scanner = new Scanner(System.in);
        while (true) {
            System.out.print("\n> ");
            if (!scanner.hasNextLine()) {
                System.out.println("\nExiting file system simulator.");
                break;
            }
            String raw = scanner.nextLine().strip();
            if (raw.isEmpty()) continue;

            String[] parts = raw.split("\\s+", 2);
            String command = parts[0].toLowerCase();
            String argument = parts.length > 1 ? parts[1].strip() : "";

            try {
                switch (command) {
                    case "format":
                        cmdFormat(freeMap, fileTable);
                        break;
                    case "create": case "read": case "write": case "delete": case "ls":
                        System.out.printf("'%s' is not yet implemented.%n", command);
                        break;
                    case "exit":
                        System.out.println("Exiting file system simulator.");
                        scanner.close();
                        return;
                    default:
                        System.out.printf("Unknown command: '%s'.%n", command);
                }
            } catch (IOException e) {
                System.out.println("Disk I/O error: " + e.getMessage());
            }
        }
        scanner.close();
    }
}